File: framework/languages/java/e2e/SpotService/Client/src/main/java/systems/zlink/e2e/spotservice/client/Scenarios/SmD15Scenario.java
package systems.zlink.e2e.spotservice.client.Scenarios;

import java.util.List;
import java.util.UUID;
import systems.zlink.e2e.spotservice.shared.Contracts;
import systems.zlink.stream.connector.ZLinkStreamConnector;

public final class SmD15Scenario extends SpotServiceScenarioContext {
    private SmD15Scenario(SpotServiceScenarioContext context) {
        super(context);
    }

    public static void run(SpotServiceScenarioContext context) {
        new SmD15Scenario(context).execute();
    }

    private void execute() {
        String actorId = "actor-sm-d15-" + UUID.randomUUID().toString().replace("-", "");
        String marker = "sm-d15-" + UUID.randomUUID().toString().replace("-", "");
        ZLinkStreamConnector connector = createStreamConnector(options().streamAEndpoint());
        try {
            Contracts.ActorProfile profile =
                new Contracts.ActorProfile("Push Chain", 15, List.of("gateway", "push"));
            connector.connect().submit().toCompletableFuture().join();
            Contracts.ActorAuthRes auth = connector
                .request(new Contracts.ActorAuthReq(actorId, profile))
                .submit(Contracts.ActorAuthRes.class).toCompletableFuture().join();
            ensure(actorId.equals(auth.actorId()), "SM-D15 auth actor mismatch");
            ensure(auth.generation() > 0, "SM-D15 auth actor generation was not concrete");

            Contracts.ActorPingRes probe = connector
                .request(new Contracts.ActorPingReq("sm-d15-bind-probe"))
                .metadata("actor-id", actorId)
                .submit(Contracts.ActorPingRes.class).toCompletableFuture().join();
            ensure(actorId.equals(probe.actorId()), "SM-D15 bind probe actor mismatch");
            ensure("play-a".equals(probe.nodeRid()), "SM-D15 bind probe node mismatch");

            var pushed = connector.waitFor(Contracts.ActorPushNotify.class)
                .where(Contracts.ActorPushNotify.class, message ->
                    actorId.equals(message.payload().actorId())
                        && marker.equals(message.payload().value()))
                .submit(Contracts.ActorPushNotify.class);
            Contracts.ActorPingRes reply = requestActorPush(actorId, marker);
            Contracts.ActorPushNotify notify = pushed.toCompletableFuture().join().payload();

            ensure(actorId.equals(reply.actorId()), "SM-D15 actor client reply actor mismatch");
            ensure("entry".equals(reply.spotRid()), "SM-D15 actor client reply spot mismatch");
            ensure("play-a".equals(reply.nodeRid()), "SM-D15 actor client reply node mismatch");
            ensure(marker.equals(reply.value()), "SM-D15 actor client reply marker mismatch");
            ensure(actorId.equals(notify.actorId()), "SM-D15 push actor mismatch");
            ensure("entry".equals(notify.spotRid()), "SM-D15 push spot mismatch");
            ensure(marker.equals(notify.value()), "SM-D15 push marker mismatch");

            waitForPlayAEvidence(List.of(
                "ActorPingReq|play-a|entry|" + actorId + "/sm-d15-bind-probe",
                "ActorPushReq|play-a|entry|" + actorId + "/" + marker));
            System.out.println("scenario SM-D15 passed");
        } catch (Exception error) {
            throw new IllegalStateException("actor push chain scenario failed", error);
        } finally {
            closeQuietly(connector);
        }

    }
}

File: framework/languages/node/e2e/SpotService/Client/Scenarios/sm-d15-scenario.ts
// SM-D15: Channel→Actor→bound Session push 사슬을 완료한다 시나리오를 검증한다.
import {
  zlinkStreamConnectorFactory,
  zlinkStreamJsonCodec,
  ZlinkStreamDispatchMode
} from '@zlink-systems/stream-connector';
import type {
  AuthReq,
  AuthRes,
  CrossRoleActorPushReq,
  CrossRoleActorPushRes,
  EvidenceWaitReq,
  ActorPushNotify,
  ActorPingReq,
  ActorPingRes
} from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { ensure } from '../Support/scenario-assert';

export async function runSmD15(options: ClientOptions): Promise<void> {
  const actorId = `actor-sm-d15-${Date.now()}`;
  const client = zlinkStreamConnectorFactory.create({
    endpoint: options.sessionAStreamEndpoint,
    codec: zlinkStreamJsonCodec,
    dispatchMode: ZlinkStreamDispatchMode.Immediate,
    heartbeat: { enabled: false },
    waitTimeoutMs: 10000
  });
  await client.connect();
  try {
    const auth = await client
      .request({
        actorId,
        displayName: actorId,
        nodeRid: 'play-a'
    } satisfies AuthReq)
      .packetName('AuthReq')
      .timeout(5000)
      .submit<AuthRes>();
    ensure(auth.nodeRid.trim().length > 0, 'SM-D15 auth reply did not include actor node rid.');
    ensure(
      auth.generation !== undefined && BigInt(auth.generation) > 0n,
      'SM-D15 auth reply did not include a positive actor generation.'
    );
    const probe = await client
      .request({ value: 'd15-bind-probe' } satisfies ActorPingReq)
      .packetName('ActorPingReq')
      .timeout(5000)
      .submit<ActorPingRes>();
    ensure(probe.actorId === actorId, 'SM-D15 bind probe actor mismatch.');

    const pushed = client.waitFor<ActorPushNotify>('ActorPushNotify')
      .timeout(10000)
      .submit();
    const reply = await postJson<CrossRoleActorPushRes>(options.playBUrl, '/actor/cross-role/push', {
      actorId,
      nodeRid: auth.nodeRid,
      generation: auth.generation,
      value: 'd15-cross-role'
    } satisfies CrossRoleActorPushReq);
    ensure(reply.delivered && reply.actorId === actorId, 'SM-D15 cross-role push reply mismatch.');
    const notify = await pushed;
    ensure(notify.payload.actorId === actorId, 'SM-D15 actor push notify actor mismatch.');
    ensure(notify.payload.value === 'd15-cross-role', 'SM-D15 actor push notify value mismatch.');
  } finally {
    await client.close();
  }

  const expected = [
    `cross-role-entry|rid=play-b|target=play-a|actor=${actorId}|value=d15-cross-role`,
    `cross-role-push|rid=play-a|actor=${actorId}|value=d15-cross-role`
  ];
  await postJson<string[]>(options.playBUrl, '/evidence/wait', {
    containsAll: [expected[0]],
    timeoutMilliseconds: 10000
  } satisfies EvidenceWaitReq);
  await postJson<string[]>(options.playAUrl, '/evidence/wait', {
    containsAll: [expected[1]],
    timeoutMilliseconds: 10000
  } satisfies EvidenceWaitReq);

  console.log('scenario SM-D15 passed');
}
